adding module sympy.matrices.random

sonntagsgesicht opened the pull request:
#### References to other Issues or PRs

#### Brief description of what is fixed or changed

- generates random matrices of various types like
  - orthogonal
  - unitary
  - invertible
  - singular
  - symmetric
  - nilpotent
  - idempotent
  - normal
  - symmetric
  - hermite 
  - elementary
  - rotation
  - reflection

#### Other comments


#### Release Notes

<!-- BEGIN RELEASE NOTES -->

* matrices
  * Added random matrix generation, i.e. for orthogonal, unitary, invertible and more matrix types.

<!-- END RELEASE NOTES -->

sonntagsgesicht committed:
adding module sympy.matrices.random
- generates random matrices of various types like
  - orthogonal
  - unitary
  - invertible
  - singular
  - symmetric
  - nilpoten
  - idempotent
  - normal

sonntagsgesicht committed:
adding module sympy.matrices.random
- generates random matrices of various types like
  - orthogonal
  - unitary
  - invertible
  - singular
  - symmetric
  - nilpoten
  - idempotent
  - normal

sonntagsgesicht committed:
adding # coding=utf-8

sonntagsgesicht committed:
updating code quality

sonntagsgesicht committed:
updating code quality

sonntagsgesicht committed:
replace pytest.raise by sympy.testing.pytest.raises

sonntagsgesicht committed:
renamed identity by identity_matrix due to conflict in combinatorics

sonntagsgesicht committed:
Merge pull request #1 from sonntagsgesicht/matrix_random

adding module sympy.matrices.random

oscargus commented on doc/src/modules/matrices/random.rst:
```diff
@@ -0,0 +1,187 @@
+
+   Even for carefully sampled matrices which are elements in topological matrix groups,
+   the sample distribution does not claim to meet any uniform distribution
+   (with respect to the groups haar measure). The sample procedures focuses
```
Haar, focus (or sample)

oscargus commented on sympy/matrices/random.py:
```diff
@@ -0,0 +1,2248 @@
+                            index=None,
+                            value=None,
+                            *scalar_set):
+    r"""super elementary matrix n x n, i.e. identity_matrix with on 2 x 2 block
```
one(?)

oscargus commented on sympy/matrices/random.py:
```diff
@@ -0,0 +1,2248 @@
+
+    Explanation
+    ===========
+    The super elementary matrix $A$ is a gerealization of elementary matrices
```
generalization

oscargus commented on sympy/matrices/random.py:
```diff
@@ -0,0 +1,2248 @@
+    * $a=0=d$ and $b=1=-c$ a ``transposition``, i.e. swapping rows $i$ and $j$
+    * $a=1$, $d=\lambda$ and $b=0=-c$ this matrix describes scaling the
+        row $j$ by $\lambda$
+    * $a=1=d$, $b=\mu, $-c=0$ adding the $\mu multiple
```
$\mu$

oscargus commented:
Very nice and valuable contribution!

Just pointed out some minor things.

I was also a bit confused with notation, but I am not sure if there is a uniform SymPy matrix notation, so I will just note it open for consideration. In general, I think we should use `\mathbf{}` for matrices. I was also surprised to see a small t for transpose (rather than capital T or \intercal) and I managed to figure out that `\bar` meant conjugate (right?). I'm used to `^H` for conjugate transpose, but then I now that it is not always used. I have not checked any other matrix documentation, so it may be that you have used exactly the same formatting. (I also thing that it is a "cultural" thing how the notations are used.)

Anyway, welcome to SymPy. Impressive first contribution!

oscarbenjamin commented on doc/src/modules/matrices/random.rst:
```diff
@@ -0,0 +1,187 @@
+
+There are three different types of matrix generators:
+
+   - the *base matrices* which from the atomic building block of the later
```
form

oscarbenjamin commented on doc/src/tutorial/matrices.rst:
```diff
@@ -231,6 +231,163 @@ filled with `0`\ s.
+Find a :ref:`full list of random matrices <matrices-random>`.
+
+    >>> import random
+    >>> random.seed(1)
```
A doctest should not modify the global random seed. Also there should be a better way to seed these functions.

oscarbenjamin replied:
See e.g. https://github.com/sympy/sympy/issues/12628

sonntagsgesicht replied:
Thx for the hint.

oscarbenjamin commented on sympy/matrices/tests/test_random.py:
```diff
@@ -0,0 +1,463 @@
+random.seed(1)
+
+
+def _check_triangular(m):
+    s, t = m.shape
+    for i in range(s):
+        for j in range(i):
+            assert m[i, j] == 0, m
```
The tests below would be clearer if this was just a boolean function e.g.:
```python
def is_triangular(m):
    s, t = m.shape
    return all(m[i, j] == 0 for i in range(s) for j in range(i))
```
Then it can be used like:
```python
assert is_triangular(M)
assert not is_triangular(N)
```

sonntagsgesicht replied:
Good point.

oscarbenjamin commented on sympy/matrices/tests/test_random.py:
```diff
@@ -0,0 +1,463 @@
+        jordan_normal(3, (None, 1, 2))
+
+
+if __name__ == '__main__':
+    import sympy
+
+    sympy.test(__file__, verbose=True, subprocess=False)
```
We shouldn't have this in a module

sonntagsgesicht replied:
Removed it.

oscarbenjamin commented on sympy/matrices/random.py:
```diff
@@ -0,0 +1,2248 @@
+
+
+def _set_value(o, i, j, v):
+    o[i, j] = v
```
What is the purpose of this function? I find the array subscript easier to understand than the function call.

sonntagsgesicht replied:
Used some of these wrappers to wrap the matrix and complex numbers framework (e.g. like `_inv` or `_re`). This keeps the code agnostic so it can be used for other frameworks (like `numpy`, `mpmath` or other.). 
I would appreciate it if we keep it that way.

oscarbenjamin replied:
Can you name a framework that doesn't use `o[i, j] =v` for setting a matrix element?

jan-philipp-hoffmann replied:
I think `tensorflow` does not support such value assignment. Well yes, `tf` doesn't support any assignment at all. 

But some naive frameworks which simply use lists of lists `[[a,  b], [c, d]]` to represent matrices would take `o[i][j]=v`. Maybe we can use `o[i, j] = v` as you prefer.

oscarbenjamin replied:
It's possible to just convert between different matrix formats so it's not clear to me what the value is in needing to support so many rather than just saying that the matrix can be constructed in one format and then converted to another.

sonntagsgesicht replied:
Since `sympy` is quite slow, creating in `sympy ` and converting it would not be efficient. Nevertheless, I already removed the `_set_value` ;)

oscarbenjamin commented on sympy/matrices/random.py:
```diff
@@ -0,0 +1,2248 @@
+    return o.inv()
+
+
+def _cs(scalar):
+    sgn = random.choice((-1, 1))
+    if isinstance(scalar, (tuple, list)) and len(scalar) == 2:
+        return scalar
+    if isinstance(scalar, complex) or getattr(scalar, 'is_real', True) is False:
+        return _re(scalar), _im(scalar)
+    c = scalar
+    s = sgn * _sqrt(1 - c * c)
+    return c, s
```
What exactly does this function do? Is it returning a random cos/sin pair?

sonntagsgesicht replied:
This functions provides proper eigenvalues of isometries. In latest version it can have real output (cos/sin) or complex (cos + sin *I). For convenience, different input types are handled (cos/sin, complex units or just cos (with random sin)). It ensures same behaviour for all isometry related matrices (like `rotation`, `reflection`, `isometry_normal` (form), `orthogonal`, `unitary` and `normal`).

oscarbenjamin commented on sympy/matrices/random.py:
```diff
@@ -0,0 +1,2248 @@
+def _cs(scalar):
+    sgn = random.choice((-1, 1))
+    if isinstance(scalar, (tuple, list)) and len(scalar) == 2:
+        return scalar
```
random.choice should not be called until it is needed

sonntagsgesicht replied:
fixed

oscarbenjamin commented on sympy/matrices/random.py:
```diff
@@ -0,0 +1,2248 @@
+
+    and $ad - bc \neq 0$.
+
+    This inculdes elementary matrices of matrix operation for Gauss elimination.
```
includes

oscarbenjamin commented on sympy/matrices/random.py:
```diff
@@ -0,0 +1,2248 @@
+
+def projection(dim,
+               index=None):
+    r"""a projection matrix n x n
```
A randomly generated n x n projection matrix

oscarbenjamin commented on sympy/matrices/random.py:
```diff
@@ -0,0 +1,2248 @@
+    Explanation
+    ===========
+    A projection is a identity_matrix like matrix
+    but with zero diagonal entiries off **index**.
```
Single `*` like `*index*` for a parameter.

Should be a blank line under `=========`

oscarbenjamin replied:
entries

oscarbenjamin commented on sympy/matrices/random.py:
```diff
@@ -0,0 +1,2248 @@
+
+
+def identity_matrix(dim):
+    r"""identity_matrix matrix n x n
```
What is the difference between this and the `eye` function.

sonntagsgesicht replied:
kind of redundancy -> removed it.

oscarbenjamin commented on sympy/matrices/random.py:
```diff
@@ -0,0 +1,2248 @@
+    If a full rank matrix $A$ is given, then
+
+    .. math::
+        B = D * (I + A * (I-D)) = D + D * A * (I-D)
```
Should this be LaTeX?

sonntagsgesicht replied:
As suggested, use now `\mathbf` for matrices in LaTeX - and `\cdot` instead of *

sonntagsgesicht commented:
Thx for the feedback and review. I'll take care of your good suggestions within the next days. Happy to adopt SymPy matrix notation.

sonntagsgesicht committed:
review suggestions
- matrices as ``\mathbf`` (doc)
- complex adjoined matrix as `\mathbf{M^H}` (doc)
- using local ``rand = random.Random()`` (matrices.random)
- removing ``identity_matrix?``(matrices.random)
- fixing typos (doc)
- _is_triangular/_is_eye/_is_zeros (test)

sonntagsgesicht committed:
flake8 fixes

sonntagsgesicht committed:
Merge pull request #2 from sonntagsgesicht/matrix_random

Added review suggestions

sonntagsgesicht commented on sympy/matrices/random.py:
```diff
@@ -0,0 +1,2248 @@
+
+        A[i,i] = a, \quad A[i,j] = b, \quad A[j,i] = -c, \quad A[j,j] = d
+
+    and $ad - bc \neq 0$.
```
yes, fixed.

sonntagsgesicht commented:
Had to re-open this as PR #22103 by my academic account. 
Code and commits are still the same. 
Sorry for the inconvenience.

oscarbenjamin commented on sympy/matrices/tests/test_random.py:
```diff
@@ -0,0 +1,535 @@
+def test_normal():
+    for d in list(TEST_DIMS)[1::3]:
+        spec = (2, 3, 4)
+        m = normal(d, spec, length=d)
+        assert all(x.is_real for x in simplify(m)), repr(simplify(m))
+        assert _is_zeros(m.T * m - m * m.T), \
+            repr(simplify(m.T * m - m * m.T))
```
With random tests like this we need to make sure that when the test fails the error message gives a way to reproduce the failure (e.g. it prints the seed or something). Otherwise if you run the tests once and it fails then you have no way to reproduce that failure afterwards.

jan-philipp-hoffmann replied:
seed is set once for all tests at the top of the test file
https://github.com/sympy/sympy/pull/22099/files/3375a9ded74f4a3350e74ea1a042c199fc18b138#r757045756
so any failure could be reproduced. 

In case of a failure, one can add some prints as done here already
https://github.com/sympy/sympy/pull/22099/files/3375a9ded74f4a3350e74ea1a042c199fc18b138#r757047379

Since there are plenty of tests (more than 600 lines of code) I don't think it would be efficient to add prints to all of them now. 

But if there are special reason to do this only for test of `normal` I think it would be feasible.

jan-philipp-hoffmann commented on sympy/matrices/tests/test_random.py:
```diff
@@ -0,0 +1,535 @@
+
+phi, psi = symbols('phi psi')
+
+rand.seed(1)
```
seed is set here

jan-philipp-hoffmann commented on sympy/matrices/tests/test_random.py:
```diff
@@ -0,0 +1,535 @@
+        assert _is_eye(i), repr(i)
+
+        m = unitary(d)
+        assert _is_eye(m.H * m, TEST_EPSILON), repr(simplify(m.H * m))
```
random test with prints
